File: leetcode/1339/solution.go
package lc1339

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func max(a, b int) int {
	if a < b {
		return b
	}
	return a
}

func getTotal(node *TreeNode) int {
	if node == nil {
		return 0
	}
	return getTotal(node.Left) + getTotal(node.Right) + node.Val
}

var total int
var product int

func getProduct(node *TreeNode) int {
	if node == nil {
		return 0
	}
	currentSum := getProduct(node.Left) + getProduct(node.Right) + node.Val
	cur_product := (total - currentSum) * currentSum
	product = max(product, cur_product)
	return currentSum
}

func maxProduct(root *TreeNode) int {

	total = getTotal(root)
	product = 0
	getProduct(root)
	return product % 1000000007
}

File: leetcode/1339/solution.py
"""
1339. Maximum Product of Splitted Binary Tree
https://leetcode.com/problems/maximum-product-of-splitted-binary-tree/
"""
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    total: int = 0
    product: int = 0

    def maxProduct(self, root: Optional[TreeNode]) -> int:
        def get_total(node: Optional[TreeNode]) -> int:
            """Recursively adds values of each node and returns total value of all nodes in the end"""
            if node is None:
                return 0
            return get_total(node.left) + get_total(node.right) + node.val

        def get_product(node: Optional[TreeNode]) -> int:
            """Recursively get the sum of nodes and update self.product if necessary"""
            if node is None:
                return 0

            # Get sum of left node, right node, and its value
            current_sum = get_product(node.left) + get_product(node.right) + node.val
            # Calculate the product
            product = current_sum * (self.total - current_sum)
            # If the product is larger than the current one, replace it
            self.product = max(self.product, product)
            # return sum of current node for future use
            return current_sum

        # Get total value of all nodes
        self.total = get_total(root)
        # Update self.product
        get_product(root)
        # Don't forget to modulo 10**9 + 7
        return self.product % 1000000007
